Samples/VirtualCamera/VirtualCameraWin32/HWMediaSource2.cpp

```diff
@@ -0,0 +1,638 @@
+
+        // Now get the event.
+        RETURN_IF_FAILED(_spEventQueue->GetEvent(dwFlags, ppEvent));
+
+        return S_OK;
+    }
+
+    IFACEMETHODIMP HWMediaSource2::QueueEvent(
+        MediaEventType eventType,
+        REFGUID guidExtendedType,
+        HRESULT hrStatus,
+        _In_opt_ PROPVARIANT const* pvValue
+    )
+    {
+        winrt::slim_lock_guard lock(m_Lock);
+
+        RETURN_IF_FAILED(_CheckShutdownRequiresLock());
+        RETURN_IF_FAILED(_spEventQueue->QueueEventParamVar(eventType, guidExtendedType, hrStatus, pvValue));
+
+        return S_OK;
+    }
+
+    // IMFMediaSource methods
+    IFACEMETHODIMP HWMediaSource2::CreatePresentationDescriptor(
+        _COM_Outptr_ IMFPresentationDescriptor** ppPresentationDescriptor)
+    {
+        winrt::slim_lock_guard lock(m_Lock);
+
+        RETURN_HR_IF_NULL(E_POINTER, ppPresentationDescriptor);
+        *ppPresentationDescriptor = nullptr;
+
+        RETURN_IF_FAILED(_CheckShutdownRequiresLock());
+        RETURN_IF_FAILED(_devSource->CreatePresentationDescriptor(ppPresentationDescriptor));
+
+        return S_OK;
+    }
+
+    IFACEMETHODIMP  HWMediaSource2::GetCharacteristics(_Out_ DWORD* pdwCharacteristics)
+    {
+        winrt::slim_lock_guard lock(m_Lock);
+
+        RETURN_IF_FAILED(_CheckShutdownRequiresLock());
+        RETURN_IF_FAILED(_devSource->GetCharacteristics(pdwCharacteristics));
+
+        return S_OK;
+    }
+
+    IFACEMETHODIMP HWMediaSource2::Pause()
+    {
+        // Pause() not required/needed
+        return MF_E_INVALID_STATE_TRANSITION;
+    }
+
+    IFACEMETHODIMP HWMediaSource2::Shutdown()
+    {
+        winrt::slim_lock_guard lock(m_Lock);
+
+        _sourceState = SourceState::Shutdown;
+
+        _spAttributes.reset();
+        if (_spEventQueue != nullptr)
+        {
+            _spEventQueue->Shutdown();
+            _spEventQueue.reset();
+        }
+
+        for (auto spStream : m_streamList)
+        {
+            spStream->Shutdown();
+        }
+        m_streamList.reset();
+        _devSource->Shutdown();
+
+        return S_OK;
+    }
+
+    IFACEMETHODIMP HWMediaSource2::Start(
+        _In_ IMFPresentationDescriptor* pPresentationDescriptor,
+        _In_opt_ const GUID* pguidTimeFormat,
+        _In_ const PROPVARIANT* pvarStartPos
+    )
+    {
+        winrt::slim_lock_guard lock(m_Lock);
+        DWORD count = 0;
+
+        wil::com_ptr_nothrow<IMFStreamDescriptor> streamDesc;
+
+        if (pPresentationDescriptor == nullptr || pvarStartPos == nullptr)
+        {
+            return E_INVALIDARG;
+        }
+        
+        if (pguidTimeFormat != nullptr && *pguidTimeFormat != GUID_NULL)
+        {
+            return MF_E_UNSUPPORTED_TIME_FORMAT;
+        }
+
+        RETURN_IF_FAILED(_CheckShutdownRequiresLock());
+
+        if (!(_sourceState != SourceState::Stopped || _sourceState != SourceState::Shutdown))
+        {
+            return MF_E_INVALID_STATE_TRANSITION;
+        }
+
+
+        // TODO: REmove

```

Comment:
>// TODO: REmove [](start = 8, length = 15)

check for TODOs throughout the solution and evaluate if should be removed / publicly commented like so